### ISSUE TITLE
How to visualize dynamically evolving networks

### ISSUE THREAD

**Issue by Fedebindi97** [projectmesa/mesa#1020]
Hey everybody!

First of all, I wanted to thank you all for the amazing work you're doing with Mesa. Keep the awesomeness!

I have an issue: I want to build a model whose network evolves at every step (i.e. new links are formed, some old links are dropped, etc.). The code _per se_ is not difficult; the problem comes with the visualization.
In fact, I don't understand how to use the NetworkGrid object to handle such a situation. The two network models in the example folder, the Boltzmann wealth model and the virus on a network model, both have fixed edges, as far as I can understand. Does NetworkGrid have methods that allow me to pop off or insert new links? From the reference page, it doesn't seem so.

Thank you in advance for the help!

**Comment by rht** [comment]
From the Boltzmann wealth model example, you can add/remove grid using standard Networkx operation (add_edge, remove_edge), operated on `self.grid.G`.

**Comment by Fedebindi97** [comment]
Thanks a lot @rht!

**Comment by Fedebindi97** [comment]
@rht, I get an error: "NetworkGrid object has no method "graph"". And furthermore, in the Boltzmann model. there is no example of the "add_edge" method being used on a grid (at least in the version that is currently uploaded in GitHub).
In general, I would like to know if it is possible to send a whole network to a grid, without having to fill the node with agents each time. I could then implement a single function making the transfer.

**Comment by rht** [comment]
`.G`, not `.graph`.

>  I could then implement a single function making the transfer.

Can you elaborate what this "transfer" is?

**Comment by Fedebindi97** [comment]
Ah ok, I thought "G" was the generic name of the graph I wanted to put into the grid.

For "transfer" I mean putting the graph entirely into the grid (with the nodes already inserted) without having to put them in one by one.

**Comment by rht** [comment]
You can't bypass doing `place_agent` one by one because inside that `place_agent` operation, the `agent.pos` needs to be informed of which node id it is located at.

Though if you meant specifically putting the graph entirely into the grid without involving agent, you can just pass in the entire graph `self.grid = NetworkGrid(self.G)`. This doesn't seem to be a problem. So not sure what you meant exactly.

**Comment by Fedebindi97** [comment]
But in the second case, doesn't it simply create a grid with empty nodes?

**Comment by rht** [comment]
Yes, is that a problem? As I said, `agent.pos` needs to know its node id, and you have to do the `place_agent` loop for each agent no matter what.

**Comment by BogdanBalcau** [comment]
Hi, so could you give an example on how to remove/add edges/nodes in a model step, please?

**Comment by rht** [comment]
@BogdanBalcau you can do this:
```python
def step(self):
    self.G.add_node(node_idx)
    self.G.remove_node(node_idx)
    self.G.add_edge(from_idx, to_idx)
    self.G.remove_edge(from_idx, to_idx)
```
Where `self.G` is your networkx object. I haven't tested this. There is a concern that `self.G` is immutable, and so the network graph passed into `self.grid` (as a `NetworkGrid`) isn't changed. But if this works, let me know. Improving the documentation for `NetworkGrid` is welcomed.

**Comment by BogdanBalcau** [comment]
> @BogdanBalcau you can do this:
> 
> ```python
> def step(self):
>     self.G.add_node(node_idx)
>     self.G.remove_node(node_idx)
>     self.G.add_edge(from_idx, to_idx)
>     self.G.remove_edge(from_idx, to_idx)
> ```
> 
> Where `self.G` is your networkx object. I haven't tested this. There is a concern that `self.G` is immutable, and so the network graph passed into `self.grid` (as a `NetworkGrid`) isn't changed. But if this works, let me know. Improving the documentation for `NetworkGrid` is welcomed.

Will try that, thank you!

**Comment by tpike3** [comment]
Closing to make issues more impactful and relevant to current Mesa issues (anything before Mesa 3.0 release Nov 9, 2024) per May 26, 2026 dev meeting.
